gapir/cc/in_memory_resource_cache.cpp

```diff
@@ -23,174 +23,237 @@
+  // to do it.
+  resize(mMemoryLimit - res.getSize());
+
+  // Get a new ID for this cache entry that is larger than any current ID.
+  auto newID = mIDGenerator++;
+
+  // Try to allocate some memory. If we get an allocation faulture, throw more
+  // stuff out until we succeed. This might happen even if we passed the memory
+  // limit check above, because we cannot control the other applications running
+  // on our device and how much memory they might use.
+  MemoryAllocator::Handle newMemory;
+  while (newMemory == nullptr) {
+    newMemory = mAllocator->allocatePurgable(res.getSize());
+
+    if (newMemory == nullptr) {
+      assert(evictLeastRecentlyUsed(mMemoryUse / 2));
+    }
   }
-  *mHead = Block(0, mBufferSize);
-  mHead->next = mHead->prev = mHead;
+
+  assert(mResourceIndex.find(res.getID()) == mResourceIndex.end());
+  assert(mResources.find(newID) == mResources.end());
+
+  // Enter the new allocation into our records.
+  mResourceIndex[res.getID()] = newID;
+  mResources[newID] = std::make_pair(res, newMemory);
+
+  // Add the memory allocated to our record of how much we have "live"
+  mMemoryUse += res.getSize();
+
+  // Copy the bits into the cache.
+  memcpy(&newMemory[0], resData, res.getSize());
+
+  return true;
 }
 
-bool InMemoryResourceCache::resize(size_t newSize) {
-  GAPID_DEBUG("Cache resizing: %zu -> %zu", mBufferSize, newSize);
-  if (newSize == mBufferSize) {
-    return true;  // No change.
-  }
+bool InMemoryResourceCache::hasCache(const Resource& res) {
+  return findCache(res) != mResources.end();
+}
 
-  Block* first = this->first();
-  Block* last = this->last();
+bool InMemoryResourceCache::loadCache(const Resource& res, void* target) {
+  mCacheAccesses++;
 
-  // Remove all the blocks that are completely beyond the end of the new size.
-  while (last != first && last->offset > newSize) {
-    last = last->prev;
-    destroy(last->next);
+  // Do we have this thing in the cache? If we don't, then we need to trigger
+  // loadCacheMiss()
+  auto resRecord = findCache(res);
+  if (resRecord == mResources.end()) {
+    if ((mCacheAccesses - mCacheHits) % 1 == 0) {
+      std::stringstream ss;
+      ss << "Replay cache miss. " << mCacheHits << " cache hits in "
+         << mCacheAccesses
+         << " accesses: " << (float)mCacheHits / (float)mCacheAccesses * 100.f
+         << " pc cache hit rate.";
+      GAPID_WARNING(ss.str().c_str());
+    }
+
+    // Get the data into the cache and return it.
+    return loadCacheMiss(res, target);
   }
 
-  if (!last->isFree()) {
-    if (last->end() > newSize) {
-      // The last block wraps the buffer. We need to evict this as we've
-      // changed the wrapping point.
-      free(last);
-    } else {
-      // Buffer has grown. Add new space block.
-      last = new Block(last->end(), 0);
-      last->linkBefore(first);
+  assert(resRecord->second.second != nullptr);  // TODO: ALAN: HANDLE THIS!!!
```

Comment:
TODO: Is this a case that we can actually get into (on purpose) or would this just indicate a bug. Either way assert() will only fire in debug builds, so we probably want something more explicit. Logging would be good.

Reply:
Debug assert that I forgot to take out. There is code down below for handling when this assert fails.